include/itkOpenCLComplexToComplex1DFFTImageFilter.hxx

```diff
@@ -0,0 +1,208 @@
+  typename OutputImageType::Pointer     outputPtr = this->GetOutput();
+
+  if (!inputPtr || !outputPtr)
+  {
+    return;
+  }
+
+  // allocate output buffer memory
+  outputPtr->SetBufferedRegion(outputPtr->GetRequestedRegion());
+  outputPtr->Allocate();
+
+  const typename InputImageType::SizeType &  inputSize = inputPtr->GetRequestedRegion().GetSize();
+  const typename OutputImageType::SizeType & outputSize = outputPtr->GetRequestedRegion().GetSize();
+
+  unsigned int vec_size = inputSize[this->m_Direction];
+  if (!this->Legaldim(vec_size))
+  {
+    ExceptionObject exception(__FILE__, __LINE__);
+    exception.SetDescription("Illegal Array DIM for FFT");
+    exception.SetLocation(ITK_LOCATION);
+    throw exception;
+  }
+
+  cl_int batchSize = 1;
+  for (unsigned int i = 0; i < TInputImage::ImageDimension; i++)
+  {
+    batchSize *= outputSize[i];
+  }
+  unsigned int totalSize = batchSize;
+  batchSize /= outputSize[this->m_Direction];
+  cl_command_queue queue = (*m_clQueue)();
+
+  if (this->m_PlanComputed) // if we've already computed a plan
+  {
+    // if the image sizes aren't the same,
+    // we have to compute the plan again
+    if (this->m_LastImageSize != totalSize)
+    {
+      delete[] this->m_InputBuffer;
+      delete[] this->m_OutputBuffer;
+      clfftDestroyPlan(&this->m_Plan);
+      this->m_PlanComputed = false;
+    }
+  }
+  if (!this->m_PlanComputed)
+  {
+    try
+    {
+      this->m_InputBuffer = new OpenCLComplexType[totalSize];
+      this->m_OutputBuffer = new OpenCLComplexType[totalSize];
+    }
+    catch (std::bad_alloc &)
+    {
+      itkExceptionMacro("Problem allocating memory for internal computations");
+    }
+    this->m_LastImageSize = totalSize;
+    const size_t n[3] = { inputSize[this->GetDirection()], 1, 1 };
+    clfftStatus  error_code = clfftCreateDefaultPlan(&this->m_Plan, (*m_clContext)(), CLFFT_1D, n);
+    if (!this->m_Plan || error_code)
+    {
+      itkExceptionMacro("Could not create OpenCL FFT Plan.");
+    }
+
+    error_code = clfftSetResultLocation(this->m_Plan, CLFFT_INPLACE);
+    error_code = clfftSetPlanBatchSize(this->m_Plan, batchSize);
+    if (std::is_same<TPixel, double>::value) // float by default
+    {
+      error_code = clfftSetPlanPrecision(this->m_Plan, CLFFT_DOUBLE);
+    }
+    clfftBakePlan(this->m_Plan, 1, &queue, nullptr, nullptr);
+    this->m_PlanComputed = true;
+  }
+
+  using InputIteratorType = itk::ImageLinearConstIteratorWithIndex<InputImageType>;
+  using OutputIteratorType = itk::ImageLinearIteratorWithIndex<OutputImageType>;
+  InputIteratorType  inputIt(inputPtr, inputPtr->GetRequestedRegion());
+  OutputIteratorType outputIt(outputPtr, outputPtr->GetRequestedRegion());
+
+  inputIt.SetDirection(this->m_Direction);
+  outputIt.SetDirection(this->m_Direction);
+
+  OpenCLComplexType * inputBufferIt = this->m_InputBuffer;
+  // for every fft line
+  for (inputIt.GoToBegin(); !inputIt.IsAtEnd(); inputIt.NextLine())
```

Comment:
In parallel?

Reply:
As the requested region is not necessarily equal to buffered region, parallelizing this is not trivial. Also, parallelizing memory access is not expected to bring big gains.